sdk/storage/storage-blob/src/IBlobSASSignatureValues.ts

```diff
@@ -161,22 +220,138 @@ export interface IBlobSASSignatureValues {
+ * WARNING: identifier will be ignored when generating user delegation SAS, permissions and expiryTime are required.
+ *
+ * @example
+ * // Generate user delegation SAS for a container
+ * const userDelegationKey = await serviceURL.getUserDelegationKey(aborter, startTime, expiryTime);
+ * const containerSAS = generateBlobSASQueryParameters({
+ *     containerName, // Required
+ *     permissions: ContainerSASPermissions.parse("racwdl").toString(), // Required
+ *     startTime, // Required. Date type
+ *     expiryTime, // Optional. Date type
+ *     ipRange: { start: "0.0.0.0", end: "255.255.255.255" }, // Optional
+ *     protocol: SASProtocol.HTTPSandHTTP, // Optional
+ *     version: "2018-11-09" // Must >= 2018-11-09 to generate user delegation SAS
+ *   },
+ *   userDelegationKey, // UserDelegationKey
+ *   accountName
+ * ).toString();
+ *
+ * @export
+ * @param {IBlobSASSignatureValues} blobSASSignatureValues
+ * @param {UserDelegationKey} userDelegationKey Return value of `ServiceURL.getUserDelegationKey()`
+ * @param {string} accountName
+ * @returns {SASQueryParameters}
+ */
+export function generateBlobSASQueryParameters(
+  blobSASSignatureValues: IBlobSASSignatureValues,
+  userDelegationKey: UserDelegationKey,
+  accountName: string
+): SASQueryParameters;
+export function generateBlobSASQueryParameters(
+  blobSASSignatureValues: IBlobSASSignatureValues,
+  sharedKeyCredentialOrUserDelegationKeyOrUserDelegationCredential:
+    | SharedKeyCredential
+    | UserDelegationKeyCredential
+    | UserDelegationKey,
+  accountName?: string
 ): SASQueryParameters {
   const version = blobSASSignatureValues.version ? blobSASSignatureValues.version : SERVICE_VERSION;
 
+  const sharedKeyCredential =
+    sharedKeyCredentialOrUserDelegationKeyOrUserDelegationCredential instanceof SharedKeyCredential
+      ? sharedKeyCredentialOrUserDelegationKeyOrUserDelegationCredential
+      : undefined;
+  let userDelegationKeyCredential =
+    sharedKeyCredentialOrUserDelegationKeyOrUserDelegationCredential instanceof
+    UserDelegationKeyCredential
+      ? sharedKeyCredentialOrUserDelegationKeyOrUserDelegationCredential
+      : undefined;
+
+  if (
+    sharedKeyCredential === undefined &&
+    userDelegationKeyCredential === undefined &&
+    accountName !== undefined
+  ) {
+    userDelegationKeyCredential = new UserDelegationKeyCredential(
+      accountName,
+      sharedKeyCredentialOrUserDelegationKeyOrUserDelegationCredential as UserDelegationKey
+    );
+  }
+
+  if (sharedKeyCredential === undefined && userDelegationKeyCredential === undefined) {
+    throw TypeError(
+      "Invalid sharedKeyCredential, userDelegationKey, accountName or userDelegationKeyCredential."
+    );
+  }
+
   // Version 2018-11-09 adds support for the signed resource and signed blob snapshot time fields.
   // https://docs.microsoft.com/en-us/rest/api/storageservices/constructing-a-service-sas#constructing-the-signature-string
   if (version >= "2018-11-09") {
```

Comment:
Suggest to first branch generate SAS or UDK SAS, and then branch version, as for UDK SAS, only version >= "2018-11-09" is valid, and different errors like throw new RangeError(
        "'version' must be >= '2018-11-09' when generating user delegation SAS using user delegation key."
      ); 
and
throw new RangeError("'version' must be >= '2015-04-05'.");

doesn't quite fit the case.

Reply:
Little different thoughts here for future maintaince aspect. Prefer the "Open–closed principle". Open to extension, close to modification.

If groupping by sharedKey & UDK like below. Then for every new API version, maintainers need to modify/add the new version into 2 groups, while the 2 groups are far away from each other. 

If we group by API version, then everytime a new API version. We just need to add a new API ersion group code, instead of modifying existing codes.

```javascript
  if (sharedKeyCredential !== undefined) {
    // Add a new version here, update existing code
    if (version >= "2018-11-09") {
    } else if (version >= "2015-04-05") {
    } else {
    }
  } else {
    // Add a new version here, update existing code
    if (version >= "2018-11-09") {
    } else {
    }
  }
```

```javascript
// Add new API version here
if (version >= "XXX") {
}

if (version >= "2018-11-09") {
}

if (version >= "2015-04-05") {
}

```

Prefer to group by API version, because we hardly to modify previous versions. The most likely secnario in the future is, a new API version.

Reply:
`throw new RangeError("'version' must be >= '2015-04-05'.");`

2015-04-05 is the minimul version required by existing SAS generation alghorithm.